All: Prevent Joplin sync ignoring new objects when syncing with an external sync service concurrently

mrjo118 opened the pull request:
When syncing to the file system sync in combination with an external sync service such as Syncthing, new files with a timestamp older than the last sync timestamp may get silently ignored (instead of being pulled into Joplin) if the external sync service syncs at the same time as the Joplin sync.

This is due to the basic delta algorithm's reliance on file timestamps, which presents a problem described well in this post https://discourse.joplinapp.org/t/photos-sometimes-not-syncing-syncthing-android-mac/47083/4:
`The main difference between filesystem sync and other types of sync is that with the latter, you've got a single source that is used by all devices, and the whole synchronisation mechanism is built around it. In the case of filesystem sync, there is no single source, but rather the filesystem sync path on each device acts as its own source. When there is just a single source, Joplin can assume that its state is always complete. However, with filesystem sync, there is no such guarantee, and this eventually leads to the issue at hand`

This change addresses the issue by allowing the remote.updated_time < context.timestamp condition of the basic delta to be bypassed if the object does not yet exist locally. Note that this has the caveat that any files in the main Joplin directory on the sync target that are not readable by Joplin will be fetched on every sync until they are manually removed. Because of this, I have made this change conditional on enabling a new 'Ignore timestamp when fetching new items' setting, with advice in the description of when it is useful to enable, and it's caveat.

Please note, this will not address the scenario where is it note updates which are ignored, however as noted in the code comment, it is possible to address this scenario without access to the original client "by making a change to the outdated note and triggering the sync. This will create a conflict containing the local note contents, and replace the main note with the latest contents from the remote version".

This partially resolves the longstanding issue https://github.com/laurent22/joplin/issues/6517. Technically the issue description only covers the new object scenario, but the issue should remain open if this change is merged, so that at some point the update object scenario may be addressed as well.

### Testing

- I have ensured that all creates, updates and fetches are cleared after 1 or 2 manual syncs, when syncing changes in both directions between 2 clients, and ensured this is the case with or without the new 'Ignore timestamp when fetching new items' setting enabled
- I have ensured that the remote.updated_time < context.timestamp condition is only bypassed when 'Ignore timestamp when fetching new items' is set to true
- I have ensured that with 'Ignore timestamp when fetching new items' is set to true, when following the reproduction steps in the description of https://github.com/laurent22/joplin/issues/6517, I am able to receive the incoming item, and I am then able to sync every future change to the item both ways between 2 clients

See video demonstrating issue demonstrating the expected outcome when the reproduction steps in the linked issue are followed:

https://github.com/user-attachments/assets/2a4f3529-9392-4f60-a191-aa6397426204

tomasz1986 commented:
> Note though, that bypassing this check has the caveat that any files in the main Joplin directory on the sync target that are not readable by Joplin, will be fetched on every sync unless they are manually removed. Because of this, I have made this change conditional upon enabling a new 'Ignore timestamp when fetching new items' setting, with advice in the description of when it is useful to enable, and it's caveat.

Is thew new setting really necessary? What do you mean exactly by files "not readable by Joplin"?

mrjo118 commented:
> What do you mean exactly by files "not readable by Joplin"?

Any files present which are not created by Joplin.
Sometimes users try to put files onto the sync target manually. Currently if they did this, after the first attempt to sync one of these files, it would get ignored by Joplin, as the context timestamp would then surpass the updated_time on the file

tomasz1986 commented:
Honestly, instead of adding the setting, wouldn't it be better to just throw an error and ask those users delete such items? I mean, as far as I am aware, there is no valid reason to fool around the sync target folder and add unrelated things in there manually. I would assume only a tiny fraction of Joplin users would do such thing anyway... or at least I hope so. 😅

mrjo118 commented:
Didn't want to mess with the existing behaviour too much, as I wouldn't want this to generate more support tickets.

To me it made sense to have no change in behaviour by default, which has a safeguard for rookie users. Then advanced users can enable this new setting, in the advanced section of sync settings, so they can have better support for Syncthing and other external sync tools. But we'll see what Laurent thinks.

tomasz1986 commented:
Thinking about this again, Syncthing also creates its own `.stfolder`/`.stversions` directories, `.stignore` file, and other temporary files when downloading data. Are these going to be a problem as well?

mrjo118 commented:
> Thinking about this again, Syncthing also creates its own `.stfolder`/`.stversions` directories, `.stignore` file, and other temporary files when downloading data. Are these going to be a problem as well?

Additional directories in the root Joplin dir are ok. But additional files will be a problem.

Possible options to exclude these external files would be to only check for files with names that are 32 chars not including the extension, or only checking files with the .md extension

tomasz1986 commented:
> only checking files with the .md extension

I think this would be ideal (because if you put your own MD files inside this particular folder, you really are asking for trouble 😅).

mrjo118 commented:
> > only checking files with the .md extension
> 
> I think this would be ideal (because if you put your own MD files inside this particular folder, you really are asking for trouble 😅).

Looks like there is an existing function which checks both that the file is 32 chars and has an md extension. That is pretty specific, so with that filtering in place I think now it does make the new setting redundant, so I've removed it.

I have updated the PR description to reflect this

mrjo118 commented:
I have now found a way to also deal with fixing updates to server objects not getting synced, which involves a toggleable alteration to the basic delta algorithm which is also backwards compatible with the old one, and with existing Joplin clients. So if this gets merged, it will fix the issue in it's entirety 😃

laurent22 commented:
Thank you for working on this but the fact that it's a rather technical setting means very few people will actually use this, yet the change brings a lot of extra complexity in terms of maintenance. I'm not sure what to do here. I guess it's not possible to enable this by default?

tomasz1986 commented:
> I guess it's not possible to enable this by default?

I also wanted to ask the same question. For clarity, I have been running Joplin with this code for about 11 days now, and everything seems to be working as expected.

@mrjo118 Is there any reason or circumstances where one would be advised not to enable this setting? Filesystem sync is broken without it, so it is basically required there. If that is indeed the case, is there any actual reason to even have an option to disable the new behaviour?

mrjo118 commented:
@laurent22 @tomasz1986 Functionally I believe that the new algorithm is sound, regardless of whether using a mirrored sync target or a single source of truth. In the case of using a hosted webdav server, the new algorithm is as good as the existing one, as the limitation around the timestamp precision also affects the existing algorithm. My concern is that enabling the new algorithm by default for everyone using file system sync and webdav is going to cause disruption for users with large data sets, because a full rescan is required on initial enablement, in order to obtain the modified timestamps of each file on file system of the server. In particular, for users of file system sync on Android, this could be a very slow process due to severe performance limitations with Android SAF introduced in Android 11+.

I do think this would be ok to be enabled by default for new instances of Joplin or new profiles though. @laurent22 what do you think about if I added some logic to default the new setting to true, but explicitly set it to false if file system / webdav sync is already setup on the initial migration of the new db schema?

tomasz1986 commented:
Just my opinion, but taking all that into consideration, I would still like to suggest to just enable it for everyone (and possibly even remove the setting) regardless of the slow scan after the upgrade (which, if I understand correctly, will be required only once).

This is because if the setting is hidden under "advanced configuration", the majority of users won't even know it exists, and the issue will still affect them (and in many cases, without them even realising that something is wrong, leading to conflicted copies or accusing Joplin and its synchronisation of being "unreliable" 😅).

mrjo118 commented:
Yes, the rescan is only required once. For file system sync, I would be open to forcing the new sync algorithm if Laurent is ok with that. However for webdav sync there is probably quite a large userbase, the majority of which will receive no gain from this change. Maybe I could enforce the new algorithm if the target is webdav and the host is localhost, otherwise use the existing algorithm?

Then either make the new setting apply to file system sync only and default to enabled for new profiles only, or remove the setting entirely and enforce the new algorithm for everyone using file system sync, depending on what @laurent22 is comfortable with

mrjo118 commented:
@laurent22 I have now updated the PR as follows:

1. I have simplified the description of the 'Detect changes based on any timestamp change' setting, changed it to only be an option for file system sync and changed it to be stored in the database, to allow setting a conditional default value at the point of migration. It is now defaulted to enabled for new profiles and Joplin installations, but if the user is currently using file system sync, when they upgrade Joplin, the setting will be set to false to avoid a full re-scan of items. I have tested that this works correctly by upgrading existing profiles with file system sync setup, WebDAV sync setup, and creating a new profile while running the new code
2. The new 'enhanced' basic delta algorithm is now used for WebDAV automatically if the url is a localhost url, otherwise WebDAV sync will use the existing basic delta algorithm (as we don't want to force a full re-scan for a feature which will not benefit those users). I have manually tested that using both a localhost url and my ip address directly (therefore not a localhost url) for a local WebDAV server correctly does or does not use the enhanced algorithm, as can be determined from the logs

If you would prefer to remove the setting altogether and force a full re-scan of items on upgrade for file system sync users, please let me know

HugoPlatzer commented:
I would really like to see this merged, because as of now, file-based sync (using Syncthing) often does not manage to sync photos in note attachments. I guess this happens mostly with large attachments and generally not with plain-text notes.

This IMHO is a pretty serious issue on an otherwise great note-taking app as sync cannot be trusted to synchronize everything - once a photo has been ignored, no clicking of the Sync button on either Android or desktop Joplin will help - only a full database rebuild causes it to recognize something is missing.

mrjo118 commented:
> @laurent22 I have now updated the PR as follows:
> 
> 1. I have simplified the description of the 'Detect changes based on any timestamp change' setting, changed it to only be an option for file system sync and changed it to be stored in the database, to allow setting a conditional default value at the point of migration. It is now defaulted to enabled for new profiles and Joplin installations, but if the user is currently using file system sync, when they upgrade Joplin, the setting will be set to false to avoid a full re-scan of items. I have tested that this works correctly by upgrading existing profiles with file system sync setup, WebDAV sync setup, and creating a new profile while running the new code
> 
> 2. The new 'enhanced' basic delta algorithm is now used for WebDAV automatically if the url is a localhost url, otherwise WebDAV sync will use the existing basic delta algorithm (as we don't want to force a full re-scan for a feature which will not benefit those users). I have manually tested that using both a localhost url and my ip address directly (therefore not a localhost url) for a local WebDAV server correctly does or does not use the enhanced algorithm, as can be determined from the logs
> 
> 
> If you would prefer to remove the setting altogether and force a full re-scan of items on upgrade for file system sync users, please let me know

@laurent22 The PR is much bigger now, but it's mostly test units in file-api.test. I've added test coverage for both the existing and new basic delta algorithm to increase confidence (there was zero test coverage for the existing algorithm, so this is an improvement). The logic for the new algorithm has not changed since the first version I implemented, which @tomasz1986 has been using for over a month now via the built binaries which I sent him

tomasz1986 commented:
I can confirm that I have been testing this PR on 4 desktop PCs and 1 Android device for about a month now, and everything seems to be working as expected (without the need for any scripts that change file timestamps into the future, etc. as is the case without this commit).

laurent22 commented:
How much of a problem is it to do a full sync for existing profiles? I wonder if we could enable it for them too since it's a one-off and anyway if someone is using filesystem sync they should probably enable the feature, even if it means having to wait a bit the first time

laurent22 commented:
(by the way there are a few conflicts)

mrjo118 commented:
> How much of a problem is it to do a full sync for existing profiles?

The only problem is them having to wait the first time, nothing else. Based on that the change is now targeted to only affect the users who will reap benefit from it, I think it makes sense to remove the configuration setting, so it is enabled for all file system sync users.

I will change this and sort out the conflicts.

mrjo118 commented:
Changes now done
